api/src/main/java/org/openmrs/module/htmlformentryui/element/ObsFromFragmentElement.java

```diff
@@ -67,10 +67,14 @@ public ObsFromFragmentElement(Map<String, String> parameters, UiUtils uiUtils) {
 		StringBuilder fragmentUrlBuilder = new StringBuilder();
 		fragmentUrlBuilder.append(fragment);
 		fragmentUrlBuilder.append("?");
-		fragmentUrlBuilder.append(parameters.get("fragmentParams"));
+        // evaluate any velocity expressions
+		String fragmentParamsString = session.evaluateVelocityExpression(parameters.get("fragmentParams"));
+		// replace any whitespace with a '+'
+		String paramsWithoutWhiteSpace = fragmentParamsString.replaceAll("\\s+", "+");
```

Comment:
@samuelmale is this meant to be URL encoding?

If yes then use the tool:
```java
java.net.URLEncoder.encode(fragmentParamsString, "UTF-8")
```
In any case, I don't think you need a new variable btw, you can just edit `fragmentParamsString`:
```java
fragmentParamsString = java.net.URLEncoder.encode(fragmentParamsString, "UTF-8");
```

Reply:
> @samuelmale is this meant to be URL encoding?

Not encoding to be more specific, just removing whitespaces.. Well I guess that's still part of the encoding activity!

> java.net.URLEncoder.encode(fragmentParamsString, "UTF-8")

Mr. tool, somehow does more than I expect it to do; asn it messing this up with our existing URL to Map parser ie.

```
@Test
	public void paramsToMap_shouldParseFragmentUrlParametersIntoAMap() throws URISyntaxException, UnsupportedEncodingException {
                 String encoded = URLEncoder.encode("path/page?age=5&gender=M", "UTF-8");
                 log.info(encoded); // path%2Fpage%3Fage%3D5%26gender%3DM
		// replay
		Map<String, Object> map = handler.paramsToMap(encoded);
		
		// verify
		Assert.assertEquals(2, map.size()); // Fails, it's actually 0
		Assert.assertEquals("{gender=M, age=5}", map.toString()); // Definitely fails too!
	}

```

Since we have a simple usecase, I would cling to using the `String#replaceAll` alongside with the strong regex!

Reply:
@samuelmale the question is whether it is ok, or even safer, to send down already encoded fragment params? Or if on the contrary we should **not** do that because of some UI Framework reason that I ignore?

As for the test, it can decode the URI before asserting, that's a detail.

Reply:
> As for the test, it can decode the URI before asserting, that's a detail.

Do you want to demonstrate that here?

> the question is whether it is ok, or even safer, to send down already encoded fragment params?

Not sure I gat your question well but all in all, method: `#paramsToMap` requires a well structure URI without whitespace..